File: _CÓDIGO FONTE/Capítulo 3/03/lista_encadeada.py
# Definição da classe Node que representa cada nó da lista encadeada
class Node:
    
    # Construtor que recebe um valor e define o próximo nó como None
    def __init__(self, valor):
        self.valor = valor
        self.proximo = None


# Definição da classe ListaEncadeada 
class ListaEncadeada:

    # Construtor que define o nó inicial como None
    def __init__(self):
        self.inicio = None

    # Método que adiciona um valor no início da lista
    def adicionar_no_inicio(self, valor):
        novo_no = Node(valor)
        novo_no.proximo = self.inicio
        self.inicio = novo_no

    # Método que adiciona um valor no final da lista
    def adicionar_no_final(self, valor):
        novo_node = Node(valor)
        if self.inicio is None:
            self.inicio = novo_node
            return
        ultimo_no = self.inicio
        while ultimo_no.proximo is not None:
            ultimo_no = ultimo_no.proximo
        ultimo_no.proximo = novo_node

    # Método que imprime todos os valores da lista
    def imprimir(self):
        no_atual = self.inicio
        while no_atual is not None:
            print(no_atual.valor)
            no_atual = no_atual.proximo


# Criação da lista encadeada e testes dos métodos
lista = ListaEncadeada()
lista.adicionar_no_final(1)
lista.adicionar_no_final(2)
lista.adicionar_no_final(3)
lista.adicionar_no_inicio(0)
lista.imprimir()

File: _CÓDIGO FONTE/Capítulo 3/03/lista_encadeada.js
class No {
    constructor(dado) {
        this.dado = dado;
        this.proximo = null;
    }
}

class ListaEncadeada {
    constructor() {
        this.cabeca = null;
    }

    adicionar(dado) {
        const novoNo = new No(dado);

        if (!this.cabeca) {
            this.cabeca = novoNo;
            return;
        }

        let atual = this.cabeca;
        while (atual.proximo) {
            atual = atual.proximo;
        }

        atual.proximo = novoNo;
    }

    inserirNoInicio(dado) {
        const novoNo = new No(dado);
        novoNo.proximo = this.cabeca;
        this.cabeca = novoNo;
    }

    inserirNaPosicao(posicao, dado) {
        const novoNo = new No(dado);
        let atual = this.cabeca;
        let anterior = null;
        let indice = 0;

        if (posicao === 0) {
            this.inserirNoInicio(dado);
            return;
        }

        while (atual && indice < posicao) {
            anterior = atual;
            atual = atual.proximo;
            indice++;
        }

        novoNo.proximo = atual;
        anterior.proximo = novoNo;
    }

    remover(dado) {
        if (!this.cabeca) {
            return;
        }

        if (this.cabeca.dado === dado) {
            this.cabeca = this.cabeca.proximo;
            return;
        }

        let atual = this.cabeca;
        let anterior = null;

        while (atual && atual.dado !== dado) {
            anterior = atual;
            atual = atual.proximo;
        }

        if (!atual) {
            return;
        }

        anterior.proximo = atual.proximo;
    }
}

// Criação da lista encadeada e testes dos métodos
var lista = new ListaEncadeada()
lista.adicionar(1) // [1]
lista.inserirNoInicio(2) // [2, 1]
lista.inserirNaPosicao(1, 3) // [2, 3, 1]
lista.remover(1) // [2, 3]
